### ISSUE TITLE
Add support for SQLite >= 3.37.0 STRICT tables

### ISSUE THREAD

**Issue by xflr6** [sqlalchemy/sqlalchemy#7398]
### Describe the use case

SQLite [release 3.37](https://www.sqlite.org/releaselog/3_37_0.html) added optional [`STRICT` tables](https://www.sqlite.org/stricttables.html) with more standard-conform type checking.

I think it would be nice to support this for the `sqlite` driver in `sqlalchemy`.

Can we include a function to change the default for `sqlite_strict` in some way (maybe on metadata)?

Maybe expose a helper for checking if the available `sqlite3` version supports strict tables?

For https://www.sqlite.org/stricttables.html#accessing_strict_tables_in_earlier_versions_of_sqlite, maybe we could provide a recipe.

### Databases / Backends / Drivers targeted

Driver: `sqlite`

### Example Use

```python
Table('some_table', metadata, ..., sqlite_strict=True)
```

**Comment by zzzeek** [comment]
of course, adding sqlite_strict is simple and we can accept PRs for this which include tests and we will also backport to the 1.4 series.  also needs some docs added to sqlite/base.py.

the hard part of STRICT will be how badly all the datatypes behave.  however, we have a long way before we start getting complaints like that as current cPython distributions won't include this latest SQLite for a little while (you can of course build python with it, but that's less common).

**Comment by zzzeek** [comment]
the `<dialect>_<option>` on the `Table` declaration is the standard approach for all parameters like this, such as `mysql_innodb`.

Standard ways to automate this currently include events like [before_parent_attach](https://docs.sqlalchemy.org/en/14/core/events.html?highlight=parent_attach#sqlalchemy.events.DDLEvents.before_parent_attach) but we don't have an example for this now, which would be suitable at https://docs.sqlalchemy.org/en/14/core/metadata.html#backend-specific-options which is right now very thin.

as for `MetaData` including parameters that are set on all tables by default, that would be a separate feature add (and also would be documented in that above section).

**Comment by FynnFreyer** [comment]
In case someone is looking for this too, and stumbles on this issue via search, like I did:

Here's a very ugly hack that gets it done (for me at least).

```python
import re

from sqlalchemy import Engine, event, Table, MetaData, Connection, text
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.schema import CreateTable


class MyBase(DeclarativeBase):
    pass


@event.listens_for(MyBase.metadata, "before_create")
def sqlite_default_strict_tables(target: MetaData, connection: Connection, **kw):
    """
    This enables strict tables by default, which is necessary to enforce column type constraints on SQLite.
    Cf. `SQLite docs <https://www.sqlite.org/stricttables.html>`
    """
    # you might have to adjust these, depending on your data types
    replacement_map = {
        r"VARCHAR(\(\d+\))?": "TEXT",
        r"DATETIME": "TEXT",
        r"DATE": "TEXT",
        r"JSON": "TEXT",
    }

    for table in target.sorted_tables:
        def redefine(target: Table, connection: Connection, **kwargs):
            if connection.engine.echo:
                print("# DROP non-strict table")
            target.drop(connection)

            statement_str = str(CreateTable(target).compile(connection)).rstrip() + " STRICT"
            for old, new in replacement_map.items():
                statement_str = re.sub(old, new, statement_str)
            statement = text(statement_str)
            connection.execute(statement)

        event.listen(table, "after_create", redefine)
```

Honestly, I don't think this is such a good idea, but if you need it, you're free to shoot your own foot with this.

**Comment by theemathas** [comment]
Here's my solution to the problem. It's based on @FynnFreyer's solution, but it should in theory be more robust (since we don't recompute the sql statement for creating the table, and I check for a few more potential errors). And it works for non-ORM use cases.

Caveats (some of which also applies to @FynnFreyer's solution):
* Requires that you already have a schema of all the tables you want to make strict.
* Untested with ORM, but replacing the redefine method in @FynnFreyer's solution with my `recreate_table_as_strict` should work.
* Untested with non-default SQLite databases (e.g. a table named `my_db.my_table`).
* All types you use that aren't [supported by SQLite strict tables](https://www.sqlite.org/stricttables.html) should be mentioned in `map_replacement_map`. None of your non-type identifiers should match these types (case-insensitive).
* Use for whatever you want but at your own risk.

```python
import re
from typing import Any

from sqlalchemy import (
    Connection,
    MetaData,
    Table,
    column,
    event,
    literal_column,
    select,
    text,
)
from sqlalchemy.dialects.sqlite import TEXT


def recreate_table_as_strict(target: Table, connection: Connection, **kwargs: Any):
    assert connection.dialect.name == "sqlite", "Only sqlite3 tables can be strict."

    table_sql_result = connection.execute(
        select(literal_column("sql", TEXT))
        .select_from(text("sqlite_schema"))
        .where(column("type") == "table")
        .where(column("name") == target.name)
    )
    table_sql = table_sql_result.fetchone()[0]  # type: ignore
    assert isinstance(table_sql, str)

    table_sql = table_sql.rstrip(" \n\t;")
    assert not table_sql.endswith("STRICT")
    table_sql = table_sql + " STRICT"

    # you might have to adjust these, depending on your data types
    type_replacement_map = {
        "DATETIME": "TEXT",
        "DATE": "TEXT",
        "TIME": "TEXT",
        "JSON": "TEXT",
    }
    for old, new in type_replacement_map.items():
        # Match only whole words, using \b in the regex
        table_sql = re.sub(r"\b" + old + r"\b", new, table_sql)

    if connection.engine.echo:
        print("# DROP non-strict table")
    target.drop(connection)
    if connection.engine.echo:
        print("# re-CREATE table as STRICT")
    connection.execute(text(table_sql))


def make_table_strict(target: Table):
    """
    Edits an existing table schema to be created as a STRICT table.
    See https://www.sqlite.org/stricttables.html
    """
    print("make_table_strict", target.name)
    event.listen(target, "after_create", recreate_table_as_strict)


def make_all_tables_strict(metadata: MetaData):
    """
    Edits an existing schema so all of its tables are created as STRICT tables.
    See https://www.sqlite.org/stricttables.html
    """
    for table in metadata.tables.values():
        make_table_strict(table)

make_all_tables_strict(put_your_metadata_here)
```

**Comment by CaselIT** [comment]
> the `<dialect>_<option>` on the `Table` declaration is the standard approach for all parameters like this, such as `mysql_engine='InnoDB'`.

having a more open ended feature similar to `prefixes` would be an option here? like `suffixes` to put stuff after the table definition. this could also be leverages for other open ended features, similar to what prefixes does?

**Comment by zzzeek** [comment]
> Here's my solution to the problem. It's based on @FynnFreyer's solution, but it should in theory be more robust (since we don't recompute the sql statement for creating the table, and I check for a few more potential errors). And it works for non-ORM use cases.


a pull request to add this keyword to SQLAlchemy would be less code and effort than these workarounds

**Comment by theemathas** [comment]
> > Here's my solution to the problem. It's based on @FynnFreyer's solution, but it should in theory be more robust (since we don't recompute the sql statement for creating the table, and I check for a few more potential errors). And it works for non-ORM use cases.
> 
> a pull request to add this keyword to SQLAlchemy would be less code and effort than these workarounds

I just found out about the [Overriding Type Compilation](https://docs.sqlalchemy.org/en/20/core/custom_types.html#overriding-type-compilation) functionality, which simplifies the most error-prone part of the workaround. I've edited my answer accordingly.

**Comment by FynnFreyer** [comment]
@theemathas looks cool, I'll test this for ORM if I get around to it.

@zzzeek I'm sure you're right, but I wouldn't even know where to start, so I use an ugly hack instead. If you could point us into the right direction, I'd consider doing it the right way

**Comment by zzzeek** [comment]
you might get an idea from this commit edef9537977 which added new table and index keywords to oracle.  that patch is bigger than this one would be.

the biggest part of the effort is getting reflection support to work, which is optional for a first pass.

**Comment by theemathas** [comment]
I think I figured it out. Here's the FINAL workaround. @FynnFreyer @zzzeek 

Running the below code will cause all tables creations in SQLite from this point on to create STRICT tables. This code is based on the supported [compilation extension mechanism](https://docs.sqlalchemy.org/en/20/core/compiler.html#changing-the-default-compilation-of-existing-constructs).

In my testing, this doesn't break indexes like my previous workaround does. I haven't tested, but I think this should work with ORM. And with some effort, it shouldn't be too hard to make this only cause some but not all tables to be STRICT.

In any case, as I mentioned before, for types you use that aren't [supported by SQLite strict tables](https://www.sqlite.org/stricttables.html), you also need to [override their compilation](https://docs.sqlalchemy.org/en/20/core/custom_types.html#overriding-type-compilation) so they get compiled into a supported SQLite type.

```python
from sqlalchemy.ext.compiler import compiles
from sqlalchemy.sql.ddl import CreateTable

@compiles(CreateTable, "sqlite")
def tables_are_strict(create_table, compiler, **kw: Any):
    return compiler.visit_create_table(create_table, **kw) + "STRICT"
```

**Comment by inklesspen** [comment]
Will it be feasible to support the [ANY](https://sqlite.org/stricttables.html#the_any_datatype) datatype, which can accept integers, reals, strings, and blobs, and returns the data in the same datatype with which it was inserted? (I suppose the Python sqlite module may have assumptions that make this difficult.)

**Comment by CaselIT** [comment]
From sqlalchemy pint of view I guess it should be possible, but that seems another issue. Feel free to one

**Comment by sqla-tester** [comment]
**Guilherme Martins Crocetti** has proposed a fix for this issue in the **main** branch:

**New SQLite DDL feature for tables: STRICT** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5610

**Comment by sqla-tester** [comment]
**Guilherme Martins Crocetti** has proposed a fix for this issue in the **rel_2_0** branch:

**SQLite strict tables** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5611
